### ISSUE TITLE
Swagger model properties exclude extra values

### ISSUE THREAD

**Issue by liviuignat** [apigee-127/swagger-tools#503]
Hi guys,

Sorry if this is a duplicate, I did an issue search and didn't find any answer. I multiple scenarios where I am sending data through `body` and reading it from `req.swagger.params.body.value` and I would expect that it read only the properties defined in the model and ignore extra properties.

So for example, if I have a model like the one bellow, I will only get those properties from the body.
```yaml
definitions:
  SomeModel:
    type: object
    required: ['name']
    properties:
      name:
        type: string
        description: The name of the company
      profilePhoto:
        type: string
        description: The profile picture path of the company
```

But actually I will receive everything that I've sent:
```json
 {
  "body": {
    "path": [
      "paths",
      "/user/me",
      "put",
      "parameters",
      "0"
    ],
    "schema": {
      "in": "body",
      "name": "body",
      "description": "The company details to change",
      "schema": {
        "type": "object",
        "required": [
          "name"
        ],
        "properties": {
          "name": {
            "type": "string",
            "description": "The name of the company"
          },
          "profilePhoto": {
            "type": "string",
            "description": "The profile picture path of the company"
          }
        }
      }
    },
    "originalValue": {
      "name": "Liviu Ignat",
      "profilePhoto": "test.jpg",
      "extraProperty": "This property should not be saved"
    },
    "value": {
      "name": "Liviu Ignat",
      "profilePhoto": "test.jpg",
      "extraProperty": "This property should not be saved"
    }
  }
}
```

Is there any feature to "crop" the parameter.value for not defined model properties?  We were using `Joi` that had this feature before and hoping to move everything to swagger models.

Thanks,
Liviu.

**Comment by liviuignat** [comment]
PS: As probably you suspect already, we are saving data to noSQL storage and we relied before on `Joi` validators to do this.

**Comment by whitlockjc** [comment]
`swagger-tools` does not work that way.  `swagger-tools` only validates your parameters and will never alter them.  One option would be to update your parameter definition to have the `additionalProperties: false` setting but I doubt that is what you want since it seems you want to send full objects but only have a subset sent to your business logic.

I don't feel it's `swagger-tools`' place to do this and so I'm closing the issue but I would love to continue the conversation in case something comes out of this that could end up in `swagger-tools`.  Long story short, if it's not Swagger related, `swagger-tools` is likely not the place.  `swagger-tools` is used for Swagger-based enforcement of routing, security and validation.  What you're looking for seems more like business logic or maybe some specialized tooling that will take a JSON Schema definition and prune out all non-explicitly described properties.

**Comment by liviuignat** [comment]
Exactly, but using an option in `swagger-tools` to do this might be handy - so you don't have duplicate code - the mapper that would prune your data would have the same structure as your schema model. 

In other languages you usually have some sort of Response/Request DTOs so you make sure you deserialise only what you need, so why to have duplicate models when they are already defined as swagger models.

Basically reading `req.swagger.params.body.value`, could give you the DTO with only what you expect from swagger to give you, if you would ever need the full `body`, you can always go to `req.body`.

**Comment by whitlockjc** [comment]
I'm not against it.  :)  I was just explaining why it was happening as-is right now.  As for how this could work, I could see a simple [vendor extension](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#vendorExtensions) that could easily identify when you want to do it and some simple function for implementing it.  Then there just needs to be some way to register vendor extension handlers to make this work.  I'd have to give this some thought because you're not registering a custom validator but a runtime processor.

**Comment by liviuignat** [comment]
Would support for custom `middleware` solve it :) ?

**Comment by whitlockjc** [comment]
Yes.  In fact, I like that better.  And it wouldn't be too hard to implement this.  The biggest problem is that `swagger-tools` is [being deprecated](https://github.com/apigee-127/swagger-tools/issues/335) so the time might be better spent getting this into `sway` _(linked to in the aforementioned link)_.  I'm cool with backporting stuff until there is a functionally equivalent middleware replacement built on top of `sway`.  Just giving you a heads up.

**Comment by liviuignat** [comment]
Was not aware that it's deprecated. Will check `sway` out. Thanks!

**Comment by whitlockjc** [comment]
Well, it's only deprecated when all functional replacements are there.  `sway` is by far a better API if you're using Swagger 2.0.  I've started the middleware but since I only work on this in my spare time, I've not made much progress on things.  Rewriting `json-refs` was a huge endeavor and since that is complete, I should have more time for `swagger-tools` and `sway`.
